Fix collapsed table row height without affecting row distribution

Ahmad-S792 opened the pull request:
#### 703a3c5807b37e2f6a79400150b5b47d486e0c89
<pre>
Fix collapsed table row height without affecting row distribution
<a href="https://bugs.webkit.org/show_bug.cgi?id=297380">https://bugs.webkit.org/show_bug.cgi?id=297380</a>
<a href="https://rdar.apple.com/158276634">rdar://158276634</a>

Reviewed by Alan Baradlay.

Previously, collapsed rows could retain nonzero height or shift subsequent rows,
causing failures in rowspan and collapsed-row cases.

This patch adds post-processe computed row heights: for each collapsed row, any
height is removed by shifting subsequent row positions up, making the collapsed
row height zero without altering other rows&apos; heights.

* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcRowLogicalHeight):
&gt; Progressions:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-002-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-003-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-group-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowcol-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowcol-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-002-border-separate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-003-border-separate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-004-dynamic-expected.txt:
* LayoutTests/TestExpectations: Unskip two tests

&gt; Rebaselines:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/other/test4-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/other/test4-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/other/test4-expected.txt:

Canonical link: <a href="https://commits.webkit.org/298783@main">https://commits.webkit.org/298783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0c66538575e231e6e71f284d3c7a42f4b312c56

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26874 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122722 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67220 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/31d04a69-9f2f-407c-a03c-9c7ba201dd68) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118537 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37010 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44901 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0eafab2a-8ec5-437e-a6cf-87308f00196c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119597 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29515 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104639 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69059 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28578 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22746 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66389 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98895 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/22904 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125858 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43547 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97259 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43911 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97052 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24711 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42375 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/20300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39509 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43433 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49028 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42899 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46239 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44605 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -329,8 +329,18 @@ LayoutUnit RenderTableSection::calcRowLogicalHeight()
     }
 
-    ASSERT(!needsLayout());
+    for (unsigned r = 0; r < totalRows; ++r) {
```
I think we prefer `size_t`

alanbaradlay replied:
use `rowIndex` maybe? r is a bit cryptic.

Ahmad-S792 replied:
Used `size_t` and also renamed to `rowIndex`.

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -329,8 +329,18 @@ LayoutUnit RenderTableSection::calcRowLogicalHeight()
-    ASSERT(!needsLayout());
+    for (unsigned r = 0; r < totalRows; ++r) {
+        if (m_grid[r].rowRenderer && m_grid[r].rowRenderer->style().visibility() == Visibility::Collapse) {
+            LayoutUnit delta = m_rowPos[r + 1] - m_rowPos[r];
```
auto?

Ahmad-S792 replied:
Done.

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -329,8 +329,18 @@ LayoutUnit RenderTableSection::calcRowLogicalHeight()
+            LayoutUnit delta = m_rowPos[r + 1] - m_rowPos[r];
+            if (delta > 0_lu) {
+                // Reduce height of collapsed row to 0 without affecting other rows
+                for (unsigned rr = r + 1; rr <= totalRows; ++rr)
```
rr?

Ahmad-S792 replied:
Renamed to `adjustedRowIndex` since it is `rowIndex -1` to handle other rows.

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -329,8 +329,18 @@ LayoutUnit RenderTableSection::calcRowLogicalHeight()
+            auto delta = m_rowPos[rowIndex + 1] - m_rowPos[rowIndex];
+            if (delta > 0_lu) {
+                // Reduce height of collapsed row to 0 without affecting other rows
+                for (unsigned adjustRowIndex = rowIndex + 1; adjustRowIndex <= totalRows; ++adjustRowIndex)
```
use `size_t` here too please

alanbaradlay replied:
`adjustedRowIndex`?

Ahmad-S792 replied:
Ops! Missed those, done both.

Ahmad-S792 committed:
Fix collapsed table row height without affecting row distribution
https://bugs.webkit.org/show_bug.cgi?id=297380
rdar://158276634

Reviewed by Alan Baradlay.

Previously, collapsed rows could retain nonzero height or shift subsequent rows,
causing failures in rowspan and collapsed-row cases.

This patch adds post-processe computed row heights: for each collapsed row, any
height is removed by shifting subsequent row positions up, making the collapsed
row height zero without altering other rows' heights.

* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcRowLogicalHeight):
> Progressions:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-002-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-003-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-row-group-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowcol-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowcol-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-002-border-separate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-003-border-separate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/visibility-collapse-rowspan-004-dynamic-expected.txt:
* LayoutTests/TestExpectations: Unskip two tests

> Rebaselines:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/other/test4-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/other/test4-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/other/test4-expected.txt:

Canonical link: https://commits.webkit.org/298783@main
